packages/firestore/src/platform_browser/webchannel_connection.ts

```diff
@@ -123,12 +123,24 @@ export class WebChannelConnection implements Connection {
                 xhr.getResponseText()
               );
               if (status > 0) {
-                reject(
-                  new FirestoreError(
-                    mapCodeFromHttpStatus(status),
-                    'Server responded with status ' + xhr.getStatusText()
-                  )
-                );
+                const response = xhr.getResponseJson().error;
+                if (!!response) {
+                  const serverError = response.status
+                    .toLowerCase()
+                    .replace('_', '-');
+                  const firestoreError: Code =
+                    Object.values(Code).indexOf(serverError) >= 0
+                      ? (serverError as Code)
+                      : Code.UNKNOWN;
```

Comment:
Can you throw this conversion code in a helper function (either in this file or in rpc_error.ts)?  mapCodeFromHttpResponse() or something?  I think giving it a name (and maybe a short comment) will make it more readable and maintainable.

`const serverError = response.status.toLowerCase().replace('_', '-');` looks a bit too arbitrary / magical out-of-context.